dbms/src/Storages/StorageKafka.cpp

```diff
@@ -0,0 +1,358 @@
+
+    if (rd_kafka_conf_set(conf, "client.id", "clickhouse", errstr, sizeof(errstr)) != RD_KAFKA_CONF_OK)
+        throw Exception(String(errstr), ErrorCodes::INCORRECT_DATA);
+
+    // Don't store offsets of messages before they're processed
+    rd_kafka_conf_set(conf, "enable.auto.offset.store", "false", nullptr, 0);
+
+    // Try to fetch preferred number of bytes before timeout
+    const Settings & settings = context.getSettingsRef();
+    auto min_bytes = settings.preferred_block_size_bytes.toString();
+    rd_kafka_conf_set(conf, "fetch.min.bytes", min_bytes.c_str(), nullptr, 0);
+}
+
+
+BlockInputStreams StorageKafka::read(
+    const Names & column_names,
+    const SelectQueryInfo & query_info,
+    const Context & context,
+    QueryProcessingStage::Enum & processed_stage,
+    size_t max_block_size,
+    unsigned num_streams)
+{
+    check(column_names);
+    processed_stage = QueryProcessingStage::FetchColumns;
+
+    if (!conf)
+    	return BlockInputStreams();
+
+    BlockInputStreams streams;
+    streams.reserve(num_streams);
+
+    // Note: The block size is set to 1, otherwise it'd have to be able to return excess buffered messages
+    for (size_t i = 0; i < num_streams; ++i)
+        streams.push_back(std::make_shared<KafkaBlockInputStream>(*this, context, schema_name, 1));
+
+    LOG_DEBUG(log, "Starting reading " << num_streams << " streams, " << max_block_size << " block size");
+    return streams;
+}
+
+
+void StorageKafka::startup()
+{
+    // Create a consumer from saved configuration
+    char errstr[512];
+    consumer = rd_kafka_new(RD_KAFKA_CONSUMER, conf, errstr, sizeof(errstr));
+    if (consumer == nullptr)
+        throw Exception("Failed to create consumer handle: " + String(errstr), ErrorCodes::UNKNOWN_EXCEPTION);
+
+    rd_kafka_poll_set_consumer(consumer);
+
+    // Create a list of partitions
+    auto * topicList = rd_kafka_topic_partition_list_new(topics.size());
+    for (const auto & t : topics)
+    {
+        LOG_TRACE(log, "Subscribing to topic: " + t);
+        rd_kafka_topic_partition_list_add(topicList, t.c_str(), RD_KAFKA_PARTITION_UA);
+    }
+
+    // Subscribe to requested topics
+    auto err = rd_kafka_subscribe(consumer, topicList);
+    if (err)
+        throw Exception("Failed to subscribe: " + String(rd_kafka_err2str(err)), ErrorCodes::UNKNOWN_EXCEPTION);
+
+    rd_kafka_topic_partition_list_destroy(topicList);
+
+    // Start the reader thread
+    stream_thread = std::thread(&StorageKafka::streamThread, this);
+}
+
+
+void StorageKafka::shutdown()
+{
+    is_cancelled = true;
+    cancel_event.set();
+
+    LOG_TRACE(log, "Unsubscribing from assignments");
+    rd_kafka_unsubscribe(consumer);
+    auto err = rd_kafka_consumer_close(consumer);
+    if (err)
+    {
+        LOG_ERROR(log, "Failed to close: " + String(rd_kafka_err2str(err)));
+    }
+
+    LOG_TRACE(log, "Destroying consumer");
+    rd_kafka_destroy(consumer);
```

Comment:
consumer is used in stream_thread, so why this code is not two lines lower?

Reply:
`rd_kafka_destroy` is asynchronous, it will close outstanding readers which will cause the `stream_thread` to terminate earlier. Then it waits for destruction completion in `rd_kafka_destroy`.